### DISCUSSION TITLE
Interactive viewer for Python bindings

### DISCUSSION THREAD

**Discussion by yuvaltassa** [google-deepmind/mujoco#576] [Announcements]
## MuJoCo Python bindings now have a built-in interactive viewer

- Usage instructions can be found in the documentation's [Python section](https://mujoco.readthedocs.io/en/latest/python.html#interactive-visualizer).
- The viewer, based on the native `simulate` library, was developed by @aftersomemath. Thanks Levi!
- It is currently **not available** from `pip` until the next official release of MuJoCo.

### Call for testers

If you you are reasonably confident in your building and packaging skills, you can build the latest version from source and help us test the new viewer. To do so, please follow the build/install instructions and report any issues you encounter on this thread.

1. `git clone https://github.com/deepmind/mujoco && cd mujoco`
2. `mkdir build && cd build`
3. `cmake -DCMAKE_BUILD_TYPE:STRING=Release -DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=${PWD}/install ..`
4. `cmake --build .`
5. `cmake --install .`
6. `cd ../python`
7. `python -m venv /tmp/mujoco_venv`
8. `source /tmp/mujoco_venv/bin/activate`
9. `./make_sdist.sh`
10. `pip install wheel`
11. `MUJOCO_PATH=${PWD}/../build/install MUJOCO_PLUGIN_PATH=${PWD}/../build/install pip wheel -v dist/mujoco-2.3.0.tar.gz`
12. `pip install --force-reinstall mujoco-2.3.0-*.whl`
13. `cd ..`

The above is for Mac and Linux. We have not tested on Windows (input welcome).

**Comment by aftersomemath** [comment]
Hi!

I just got to test the version on `main`. Almost everything worked as expected. :)

Two issues of general interest that I found are:
* If you use `simulate.load(model)` or `python -m mujoco.simulate --mjcf=/path/to/some/mjcf.xml` to start the UI, then using the reload button causes the physics loop to crash since `dropfilename` is not set
* There is no programmatic or command line way to set an XML filename for reloading, only drag-n-drop can do it
* `dropfilename` is being used instead of `filename` when the UI load button is pressed (this is just a typo)

**Possible solution:** Add an optional argument to `simulate.load` that accepts a string to an XML file. In the case that the user is passing `model` and/or `data` objects, add arguments to `simulate.load` for user defined callbacks that provide the `model` and `data` objects. It may be appropriate to have two `launch` methods. `launch` for `model` and `data` launches and `launch_xml_path` for launching with a path to an XML file.

One issue specific to my use case is:
* If the reload button is "fixed" (which I did by hardcoding `dropfilename` to a XML file's path in `simulate.h`) and `mjcb_control` is set, then clicking reload causes an exception somewhere in the C++ to Python callback mechanism.

**Possible solution**: I believe the problem is that `mjcb_control` needs to be set after the new model object is created so that the C++ pointer to Python wrapper lookup table is updated. If the callbacks that were proposed above are implemented then the user would be able to use the callbacks to reset `mjcb_control` and any other callbacks appropriately (they may also do more complicated things like save recordings to disk, reset states external to MuJoCo, etc).

I am happy to implement these changes if the team thinks they make sense.

**Reply by saran-t** [reply]
I think https://github.com/deepmind/mujoco/commit/9fcf3b0c2e7f4d089d188a7a5ea183ed66b6bd35 resolves a lot of these issues, thanks for pointing them out!

I haven't looked at `mjcb_control` yet but I suspect whatever the problem was it's to do with your specific implementation of it. Try out the new commit and maybe point me to your `mjcb_control` implementation?

**Reply by aftersomemath** [reply]
Thanks! `9fcf3b0` resolved all of the issues except for the `mjcb_control` issue. Sorry for the delay in responding.

[Here](https://github.com/aftersomemath/mujoco/blob/e1ba2e6e2220758539f034b5a5ddbcdfdb4c6e34/model/double_pendulum/double_pendulum.py#L56) is a short example of the `mjcb_control` issue. As written, the example works. But, if the `mujoco.set_mjcb_control(None)` call is removed from the very beginning of the `load_callback` then an exception is raised when the reload button is pressed.

The exception occurs even if `mjcb_control` is set to `def double_pendulum_control(m, d): pass`, so I think it is not the callback but something to do with the `MjWrapperLookup` failing. But, I haven't looked into it thoroughly.

**Comment by vikashplus** [comment]
I was able to get it to work on my pop!-os 20.04 with a small workaround - 
Step3 errored out with `RandR headers not found; install libxrandr development package`. I did `sudo apt-get install xorg-dev libglu1-mesa-dev` following suggestion from [here](https://github.com/openMVG/openMVG/issues/85#issuecomment-34821174)

**Reply by worthy-wire** [reply]
Had the same error on Ubuntu 22.04.1 and this worked for me as well

**Reply by saran-t** [reply]
As an alternative, the minimal dependency set that we use for our GitHub Actions CI build is [here](https://github.com/deepmind/mujoco/blob/main/.github/workflows/build.yml#L105).

**Comment by vikashplus** [comment]
Following gives me an empty visualization with staring screen prompt (Drag and drop to load model). 
Am I using it the correct way? @yuvaltassa @saran-t ?

```
import mujoco
from mujoco import viewer

sim_path = 'model/humanoid/humanoid.xml'
m = mujoco.MjModel.from_xml_path(sim_path)
d = mujoco.MjData(m)

viewer.launch_repl(m, d)
```
mujoco==2.3.1.post1
dm-control==1.0.9
mac-os | Monterey | v12.6.1

**Comment by saran-t** [comment]
`launch_repl` at 2.3.1.post1 is buggy. We've fixed it at head but probably
won't get around to releasing a prebuilt for another couple of weeks.

**Reply by yuvaltassa** [reply]
@vikashplus can you build from source and let us know if you still experience any issues?

**Reply by vikashplus** [reply]
Im traveling for next two weeks. I'll let you know if I get around it.

**Comment by matinmoezzi** [comment]
Hi, I am trying to move a robotic arm in Mujoco using the keyboard keys like what has been implemented in `mujoco_py.mjviewer`. My question is how to implement functions like `add_overlay` or `key_callback` to this interactive viewer for Python bindings (`mujoco.viewer`). This repo [https://github.com/rohanpsingh/mujoco-python-viewer](https://github.com/rohanpsingh/mujoco-python-viewer) has implemented such viewer using Mujoco Python binding. I am wondering if there is such a viewer built-in in the interactive viewer (`viewer.launch()`).
